### PR TITLE
Final Fix 6146: princess bombing flying infantry

### LINKED ISSUE #6146: Bombs, elevation, and Princess
### Prerequisites and Pre-Issue Checklist

- [X] I'm reporting the issue to the correct repository:
- [MegaMek](https://github.com/MegaMek/megamek/issues)
- [MegaMekLab](https://github.com/MegaMek/megameklab/issues)
- [MekHQ](https://github.com/MegaMek/mekhq/issues)

- [X] I've tested the issue against at least the latest MILESTONE version
- [X] I've asked on the MegaMek Discord about the error
- [X] I've reviewed the BattleTech rules and MegaMek documentation, and I've confirmed that something isn't working as intended.
- [X] I've searched the Github tracker and haven't found the issue listed

### Severity *

Medium (Gameplay Limitation): Non-core functionality is impaired, providing a suboptimal but playable experience.

### Brief Description *

When facing infantry which has vertical movement (such as branth beast infantry), Princess seems to only see them as infantry and will attempt to bomb them with ASF units. As the infantry is at an elevation the bombs are unable to do damage even if they hit the correct hex, but Princess will happily but ineffectively drop the payload anyway. She should probably be able to tell she's going to do zero damage.

Attached save for the turn of combat where it happens, as well as screencaps of the round report showing the bombing attempt.

### Steps to Reproduce

This is reproducible in any situation where infantry exists at a above ground elevation versus bombing aircraft. 

### Operating System *

Windows 11

### Java Version *

17

### MegaMek Suite Version *

v0.50.0

### Custom MegaMek Version

_No response_

### Attach Files

Untitled
Untitled2
[HE.sav.gz](https://github.com/user-attachments/files/17531632/HE.sav.gz)

### Final Checklist

- [X] I've checked to make sure that this issue has not already been filed
- [X] I'm reporting only one issue in this ticket for clarity and focus

### PR DESCRIPTION
**Pull request by Sleet01**
This PR constitutes a major rework of AE blast handling and Princess bombing.

### Issues With Existing Bombing and Artillery Handling ###
During the initial work to fix #6146 I identified some differences between our AE blast damage implementation and the rules per Total Warfare and Tactical Operations:

1. AE attacks that hit buildings and water are supposed to generate a blast "sphere"; we only create a disk.
2. Artillery shots are supposed to create a "top" shape, disk + spindle, but we only create the disk.
3. Artillery shots that hit buildings are supposed to be handled per 1); we just do 2).
4. Bombs should be able to target specific levels of buildings; we just target the ground.
5. Flak Artillery shots at VTOLs should create a blast "top" shape, with a spindle up and down and the normal blast disk; we don't do that.
6. Princess should be able to target VTOL and UMU/Submarine units within 1 (or 2) levels of Building tops and Water surface levels; initially she targeted any unit including Aerospace and high-flying VTOL units.
7. Princess should spread bombing damage out; currently she concentrates all bomb attacks on a single hex or target unit.
8. Princess doesn't evaluate bombing plans based on enemy damage and friendly damage; instead, she just _subtracts_ friendly damage from enemy damage, which exacerbates 7) above.
9. Non-Flak blast damage should be capable of hitting low-flying VTOLs per above; we were skipping VTOLs completely when applying Artillery and Bomb damage, even if they would have fallen within the existing blast area calculations.

### Fixes and Updates ###

The fixes consist of three major changes:
1. I started by creating a method to build a "3D" blast shape according to RAW, encoding AE base damage at specific level-hex coordinates, per TW and TacOps.  This allows dealing damage to various target types just by iterating through the blast "shape" entries in order.

2. Overhauled much of the way that Princess estimates bombing and artillery damage, along with cleaning up some of the expected damage code and variable names.  This includes new blast damage estimation code, similar to the code for 1) above, but faster (if slightly less accurate).
It is now possible to evaluate building and friendly fire damage separately from damage to enemies, which should allow for finer aggression tuning.
Princess will also _try_ to avoid massively overkilling a single target, including infantry, with concentrated bomb attacks.

4. Began work on consolidating all blast radius info into `AreaEffectHelper.calculateDamageFallOff()`.  
Initially I was considering putting the blast radius info into AmmoType, similar to how some (but not all!) BombTypes were doing, but that actually would have been more work, and anyhow `calculateDamageFallOff()` was already being used in a lot of places.

These three major pieces of work should fix the issues identified above without massively increasing processing time.

### Still to Do ###

There is still some work needed to make MegaMek fully TW / TacOps-compliant:

1. Add targetLevel field to information passed over the wire when sending attack packets from client to server.  This is a relatively small change but non-trivial.
2. Provide mechanism for players to specify a target level when bombing buildings; this is expressly described in Total Warfare.

There are also some updates to Princess that would allow her to more effectively utilize bombs:
1. Allow adjusting Dive Bomb attacks to use fewer bombs if the chance of overkill is very high; we already do something similar to bring normal firing plans under heat limits.
2. Teach Princess to use level bombing; the new functions to estimate blast damage should help here.

### Testing ###

I conducted numerous manual, semi-manual, and bot-on-bot tests to validate this code, but stress testing is definitely needed to find edge cases that need better handling.

I did add unit tests for some of the new code, as well as updating existing tests to use the new method of reporting expected damage.

### CODE REVIEW

File: megamek/src/megamek/common/weapons/AreaEffectHelper.java [hunk 1]
```diff
@@ -394,29 +399,19 @@ public static void artilleryDamageEntity(Entity entity, int damage, Building bld
             return;
```
Comment (rjhancock) [inline megamek/src/megamek/common/weapons/AreaEffectHelper.java:399]: Up about 30 LOC there seems a possible null dereferencing in regards to `hex` as it's checked somewhere within 100 LOC below this line. Right after the comment about `//check: is entity inside building?`

File: megamek/unittests/megamek/client/bot/princess/WeaponFireInfoTest.java [hunk 2]
```diff
@@ -339,4 +339,10 @@ void testPostProcessingHomingNoTAGAround() {
         assertEquals(expectedTHD.getDesc(), processed.getDesc());
         assertNotEquals(originalTHD, processed);
     }
+
+    @Test
+    void testComputeExpectedBombDamageOnlyEnemy() {
+//        WeaponFireInfo testBombWFI = setupBombWFI();
+
+    }
```
Comment (rjhancock) [inline megamek/unittests/megamek/client/bot/princess/WeaponFireInfoTest.java:347]: No need for an empty test. :)

File: megamek/unittests/megamek/common/weapons/AreaEffectHelperTest.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
+package megamek.common.weapons;
+
+import megamek.client.Client;
+import megamek.client.ui.swing.ClientGUI;
+import megamek.common.*;
```
Comment (rjhancock) [inline megamek/unittests/megamek/common/weapons/AreaEffectHelperTest.java:5]: Copyright...

File: megamek/unittests/megamek/common/ComputeTest.java [hunk 4]
```diff
@@ -11,6 +11,8 @@
 import org.junit.jupiter.api.BeforeEach;
```
Comment (rjhancock) [inline megamek/unittests/megamek/common/ComputeTest.java:11]: Copyright...

**Review by rjhancock — COMMENTED** [review]
The copyrights are the only thing I insist on. Everything else are nitpicks.

**Comment by Sleet01** [comment]
Not going to have time to work on this further until tomorrow.
To do:
1. fix above nits
2. add copyright blocks
3. plumb FAE blasts through the new code

File: megamek/unittests/megamek/client/bot/princess/WeaponFireInfoTest.java (hunk 2 above)
Comment (Sleet01) [reply megamek/unittests/megamek/client/bot/princess/WeaponFireInfoTest.java:347]: Dangit.

File: megamek/src/megamek/common/weapons/AreaEffectHelper.java (hunk 1 above)
Comment (rjhancock) [reply megamek/src/megamek/common/weapons/AreaEffectHelper.java:399]: Always edge cases can derail us so good to be safe. :)
